### PR TITLE
APIs and strawman generation for exposing .NET objects to COM

### PR DESCRIPTION
**Pull request by jkoritzinsky**
## Design

We introduce a `GeneratedComClassAttribute`. This attribute is applied to a `StrategyBasedComWrappers`-derived type once per each type the user wants to expose to COM. We then generate code to compute the vtables based on the compile-time interface list of the type, using the provided `IIUnknownInterfaceDetailsStrategy` to ensure comparable behavior between CCW and RCW scenarios (an interface is considered a COM interface for either both or neither scenario).

In `ComputeVtables`, we determine the type of the passed-in object and return the corresponding set of vtables.

### Options

In the source generator, we can detect if the user has overridden the `GetOrCreateInterfaceDetailsStrategy` method. We could optimize the code we emit when it is using the base implementation (that uses the default strategy) to calculate the vtables statically (instead of per-ComWrappers instance) as we know the default implementation is a singleton.

We could also not support custom strategies for the CCW side, but that feels like it would just cause customer confusion as rules for getting interface details would differ between the RCW and CCW sides of the generator design.

## Notes

Includes #7.

Updates the example to use the CCW APIs as the "COM" side of the implementation instead of the manually written COM objects. I added an assert to make sure that we don't accidentally miss testing the RCW side.

### CODE REVIEW

File: Generated/MyGeneratedComWrappers.cs [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+{
+    private ComInterfaceEntry* s_vtableA = null;
+    private ComInterfaceEntry* VTableA
+    {
+        get
+        {
+            if (s_vtableA != null)
+                return s_vtableA;
+            IIUnknownInterfaceDetailsStrategy detailsStrategy = GetOrCreateInterfaceDetailsStrategy();
+            ComInterfaceEntry* vtable = (ComInterfaceEntry*)RuntimeHelpers.AllocateTypeAssociatedMemory(typeof(A), sizeof(ComInterfaceEntry) * 1);
```
Comment (jtschuster) [inline Generated/MyGeneratedComWrappers.cs:15]: Why is this not `sizeof(ComInterfaceEntry) * 4`, with 3 for IUnknown + 1?

File: Generated/IComInterface1.cs [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+{
+	public static void GetIUnknownImpl(out void* pQueryInterface, out void* pAddRef, out void* pRelease)
+    {
+		nint qi, addRef, release;
+        ComWrappers.GetIUnknownImpl(out qi, out addRef, out release);
+        pQueryInterface = (void*)qi;
+        pAddRef = (void*)addRef;
+        pRelease = (void*)release;
+    }
+}
```
Comment (AaronRobinsonMSFT) [inline Generated/IComInterface1.cs:15]: These are no longer needed in .NET 8, right? I want to make sure we aren't going to be doing this in the actual source generator.

File: Generated/IComInterface1.cs [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        var (thisPtr, vtable) = ((IUnmanagedVirtualMethodTableProvider)this).GetVirtualMethodTableInfoForKey(typeof(global::IComInterface1));
+        int hr = ((delegate* unmanaged<void*, int>)vtable[3])(thisPtr);
+        if (hr < 0)
+        {
+            Marshal.ThrowExceptionForHR(hr);
+        }
+    }
+}
+
+[IUnknownDerived<InterfaceInformation, InterfaceImplementation>]
```
Comment (AaronRobinsonMSFT) [inline Generated/IComInterface1.cs:73]: Based on this, it seems that from the C# perspective the `file` scope means the type is only visible within this translation unit. However we can still pass the type handle around and use/discover it via the attribute, right? From outside this scope it is merely an opaque Type handle? If so, can I assume the `file` scope is entirely C# semantics and has no influence over the actual IL?

Comment (AaronRobinsonMSFT) [reply Generated/IComInterface1.cs:73]: Ugh. Nope. I defined `file class A {}` in SharpLabs and the resulting type name was `<_>FD2E2ADF7177B7A8AFDDBC12D1634CF23EA1A71020F6A1308070A16400FB68FDE__A`. This is going to make diagnostics incredibly annoying.

File: Generated/IComInterface1.cs (hunk 2 above)
Comment (jkoritzinsky) [reply Generated/IComInterface1.cs:15]: Yes, this is no longer needed in .NET 8.

File: Generated/IComInterface1.cs (hunk 3 above)
Comment (jkoritzinsky) [reply Generated/IComInterface1.cs:73]: Yeah for putting the types into metadata, there's not a good way to have multiple entries with the same name, so they decided to do a prefix with the hash of something.

Comment (jtschuster) [reply Generated/IComInterface1.cs:73]: FWIW it looks liike they're named `<FileName>HASH__TypeName`. Still annoying, but with how we separate code for each generated type into its own file it's not the worst.

Comment (AaronRobinsonMSFT) [reply Generated/IComInterface1.cs:73]: This issue is with things like memory allocation tools that show types and stack analysis that show frames. The method name is going to have a really annoying class name and the allocated type is rather verbose. I've contacted the appropriate VS teams, they are going to need to think about an "pretty printing" transform.

**Comment by jkoritzinsky** [comment]
Closing at #8 supersedes this design.

**Pull request closed without merging**